homeassistant/components/logbook/websocket_api.py

```diff
@@ -0,0 +1,304 @@
+from homeassistant.components.websocket_api import messages
+from homeassistant.components.websocket_api.connection import ActiveConnection
+from homeassistant.components.websocket_api.const import JSON_DUMP
+from homeassistant.core import CALLBACK_TYPE, Event, HomeAssistant, callback
+import homeassistant.util.dt as dt_util
+
+from .helpers import (
+    async_determine_event_types,
+    async_filter_entities,
+    async_subscribe_events,
+)
+from .models import async_event_to_row
+from .processor import EventProcessor
+
+MAX_PENDING_LOGBOOK_EVENTS = 2048
+EVENT_COALESCE_TIME = 0.5
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@callback
+def async_setup(hass: HomeAssistant) -> None:
+    """Set up the logbook websocket API."""
+    websocket_api.async_register_command(hass, ws_get_events)
+    websocket_api.async_register_command(hass, ws_event_stream)
+
+
+async def _async_get_ws_formatted_events(
+    hass: HomeAssistant,
+    msg_id: int,
+    start_time: dt,
+    end_time: dt,
+    formatter: Callable[[int, Any], dict[str, Any]],
+    event_processor: EventProcessor,
+) -> tuple[str, dt | None]:
+    """Async wrapper around _ws_formatted_get_events."""
+    return await get_instance(hass).async_add_executor_job(
+        _ws_formatted_get_events,
+        msg_id,
+        start_time,
+        end_time,
+        formatter,
+        event_processor,
+    )
+
+
+def _ws_formatted_get_events(
+    msg_id: int,
+    start_day: dt,
+    end_day: dt,
+    formatter: Callable[[int, Any], dict[str, Any]],
+    event_processor: EventProcessor,
+) -> tuple[str, dt | None]:
+    """Fetch events and convert them to json in the executor."""
+    events = event_processor.get_events(start_day, end_day)
+    last_time = None
+    if events:
+        last_time = dt_util.utc_from_timestamp(events[-1]["when"])
+    result = formatter(msg_id, events)
+    return JSON_DUMP(result), last_time
+
+
+async def _async_events_consumer(
+    setup_complete_future: asyncio.Future[dt],
+    connection: ActiveConnection,
+    msg_id: int,
+    stream_queue: asyncio.Queue[Event],
+    event_processor: EventProcessor,
+) -> None:
+    """Stream events from the queue."""
+    subscriptions_setup_complete_time = await setup_complete_future
+    event_processor.switch_to_live()
+
+    while True:
+        events: list[Event] = [await stream_queue.get()]
+        # If the event is older than the last db
+        # event we already sent it so we skip it.
+        if events[0].time_fired > subscriptions_setup_complete_time:
+            # We sleep for the EVENT_COALESCE_TIME so
+            # we can group events together to minimize
+            # the number of websocket messages when the
+            # system is overloaded with an event storm
+            await asyncio.sleep(EVENT_COALESCE_TIME)
+            while True:
+                try:
+                    events.append(stream_queue.get_nowait())
+                except asyncio.QueueEmpty:
+                    break
```

Comment:
This seems more efficient than going into try…catch inside a loop
```suggestion
            while not stream_queue.empty():
                events.append(stream_queue.get_nowait())
```